src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -3350,6 +3432,321 @@ public <T> T invokeAny(Collection<? extends Callable<T>> tasks,
             .invokeAny(tasks, this, true, unit.toNanos(timeout));
     }
 
+    // Support for delayed tasks
+
+    /**
+     * Returns STOP and SHUTDOWN status (zero if neither), masking or
+     * truncating out other bits.
+     */
+    final int shutdownStatus(DelayScheduler ds) {
+        return (int)(runState & (SHUTDOWN | STOP));
+    }
+
+    /**
+     * Tries to stop and possibly terminate if already enabled, return success.
+     */
+    final boolean tryStopIfShutdown(DelayScheduler ds) {
+        return (tryTerminate(false, false) & STOP) != 0L;
+    }
+
+    /**
+     *  Creates and starts DelayScheduler
+     */
+    private DelayScheduler startDelayScheduler() {
+        DelayScheduler ds;
+        if ((ds = delayScheduler) == null) {
+            boolean start = false;
+            String name = poolName + "-delayScheduler";
+            if (workerNamePrefix == null)
+                asyncCommonPool();  // override common parallelism zero
+            lockRunState();
+            try {
+                if ((ds = delayScheduler) == null) {
+                    ds = delayScheduler = new DelayScheduler(this, name);
+                    start = true;
+                }
+            } finally {
+                unlockRunState();
+            }
+            if (start) { // start outside of lock
+                // exceptions on start passed to (external) callers
+                SharedThreadContainer ctr;
+                if ((ctr = container) != null)
+                    ctr.start(ds);
+                else
+                    ds.start();
+            }
+        }
+        return ds;
+    }
+
+    /**
+     * Arranges execution of a ScheduledForkJoinTask whose delay has
+     * elapsed
+     */
+    final void executeEnabledScheduledTask(ScheduledForkJoinTask<?> task) {
+        externalSubmissionQueue(false).push(task, this, false);
+    }
+
+    /**
+     * Arranges delayed execution of a ScheduledForkJoinTask via the
+     * DelayScheduler, creating and starting it if necessary.
+     * @return the task
+     */
+    final <T> ScheduledForkJoinTask<T> scheduleDelayedTask(ScheduledForkJoinTask<T> task) {
+        DelayScheduler ds;
+        if (((ds = delayScheduler) == null &&
+             (ds = startDelayScheduler()) == null) ||
+            (runState & SHUTDOWN) != 0L)
+            throw new RejectedExecutionException();
+        ds.pend(task);
+        return task;
+    }
+
+    /**
+     * Submits a one-shot task that becomes enabled after the given
```

Comment:
If the javadoc for the new methods from the ScheduledExecutorService interface are cut-and-pasted from that interface, then I recommend using the directed inherit doc feature which would reduce the javadoc text in this class to something like:

```
/**
 * {@inheritdoc ScheduledExecutorService }
 * @param command   {@inheritdoc ScheduledExecutorService }
 * ...
 * @throws NullPointerException   {@inheritdoc ScheduledExecutorService }
 */
```

Reply:
The schedule(Runnable ..) and schedule(Callable ..) methods are a single sentence in the interface. The ForkJoinPool methods has more to say about cancellation and shutdown. So while it could inherit, doesn't seem to be worth it here.

The descriptions in scheduleAtFixedRate and scheduleWithFixedDelay aren't identity to the interface because there are additional exceptional completions to list. So I don't think these can inherit.

Reply:
Agreed. The root cause is that ScheduledExecutorService should have mentioned interactions with shutdown and cancellation, but didn't so both FJP and ScheduledThreadPoolExecutor had to add them (almost identically).

Reply:
Just to confirm: The current version seems preferable to alternatives